File: server/db/migrate.ts
import 'dotenv/config'
import { readdir, readFile } from 'node:fs/promises'
import { fileURLToPath } from 'node:url'
import path from 'node:path'
import { pool } from './pool'

const currentDirectory = path.dirname(fileURLToPath(import.meta.url))
const migrationsDirectory = path.join(currentDirectory, 'migrations')

async function migrate() {
  const client = await pool.connect()
  try {
    await client.query(`
      CREATE TABLE IF NOT EXISTS schema_migrations (
        name text PRIMARY KEY,
        applied_at timestamptz NOT NULL DEFAULT now()
      )
    `)
    const files = (await readdir(migrationsDirectory))
      .filter((file) => file.endsWith('.sql'))
      .sort()

    for (const file of files) {
      const existing = await client.query('SELECT 1 FROM schema_migrations WHERE name = $1', [file])
      if (existing.rowCount) continue
      const sql = await readFile(path.join(migrationsDirectory, file), 'utf8')
      await client.query('BEGIN')
      try {
        await client.query(sql)
        await client.query('INSERT INTO schema_migrations (name) VALUES ($1)', [file])
        await client.query('COMMIT')
        console.log(`[database] applied ${file}`)
      } catch (error) {
        await client.query('ROLLBACK')
        throw error
      }
    }
  } finally {
    client.release()
    await pool.end()
  }
}

migrate().catch((error) => {
  console.error('[database] migration failed', error)
  process.exitCode = 1
})

File: backend/app/migrate.py
from pathlib import Path
import os

from dotenv import load_dotenv
import psycopg


def migrate() -> None:
    project_root = Path(__file__).resolve().parents[2]
    load_dotenv(project_root / ".env")
    database_url = os.environ.get("DATABASE_URL")
    if not database_url:
        raise RuntimeError("DATABASE_URL is required before applying database migrations")
    migrations = project_root / "database" / "migrations"
    with psycopg.connect(database_url) as conn:
        conn.execute(
            "CREATE TABLE IF NOT EXISTS schema_migrations (name text PRIMARY KEY, applied_at timestamptz NOT NULL DEFAULT now())"
        )
        for file in sorted(migrations.glob("*.sql")):
            exists = conn.execute("SELECT 1 FROM schema_migrations WHERE name=%s", (file.name,)).fetchone()
            if exists:
                continue
            sql = file.read_text(encoding="utf-8")
            with conn.transaction():
                conn.execute(sql, prepare=False)
                conn.execute("INSERT INTO schema_migrations (name) VALUES (%s)", (file.name,))
            print(f"[database] applied {file.name}")


if __name__ == "__main__":
    migrate()
